library/foreman_provisioner.py

```diff
@@ -0,0 +1,292 @@
+        finished rebuilding before exiting.
+        :raises: KeyError if BMC hasn't been found on the given host
+                 Exception in case of machine could not be reached after
+                 rebuild
+        """
+        if not self._validate_bmc(host_id):
+            raise KeyError("BMC not found on {}".format(host_id))
+        building_host = self.get_host(host_id)
+        self.set_build_on_host(host_id, True)
+        if mgmt_strategy == 'foreman':
+            self.bmc(host_id, mgmt_action)
+        elif mgmt_strategy == 'ipmi':
+            host = "{0}.{1}".format(building_host.get('interfaces')[0].get('name'), building_host.get('domain_name'))
+            self.ipmi(host, mgmt_action)
+        else:
+            raise Exception("{0} is not a supported "
+                            "management strategy".format(mgmt_strategy))
+        if wait_for_host:
+            while self.get_host(host_id).get('build'):
+                time.sleep(WAIT_TO_FINISH_BUILDING)
+
+            command = "ping -q -c 30 -w 300 {0}".format(building_host.get('name'))
+            return_code = subprocess.call(command.split(" "))
+
+            if return_code > 0:
+                raise Exception("Could not reach {0}, rc={1}, cmd={2}".format(host_id, return_code, command))
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            username=dict(default='admin'),
+            password=dict(required=True),
+            auth_url=dict(required=True),
+            host_id=dict(required=True),
+            rebuild=dict(default=True, choices=BOOLEANS),
+            mgmt_strategy=dict(default='foreman',
+                               choices=MGMT_SUPPORTED_STRATEGIES),
+            mgmt_action=dict(default='cycle', choices=['on', 'off', 'cycle',
+                                                       'reset', 'soft']),
+            wait_for_host=dict(default=True, choices=BOOLEANS)))
+
+    foreman_client = ForemanManager(url=module.params['auth_url'],
+                                    username=module.params['username'],
+                                    password=module.params['password'])
+
+    status_changed = False
+
+    if module.boolean(module.params['rebuild']):
+        try:
+            foreman_client.provision(module.params['host_id'],
+                                     module.params['mgmt_strategy'],
+                                     module.params['mgmt_action'],
+                                     module.boolean(
+                                         module.params['wait_for_host']))
+        except KeyError as e:
+            module.fail_json(msg=e.message)
+        else:
+            status_changed = True
+
+    #TODO(tkammer): implement RESERVE and RELEASE
+    host = foreman_client.get_host(module.params['host_id'])
+    interface = foreman_client.get_host('{0}/interfaces'.format(module.params['host_id']))
+    if host.has_key('error'):
```

Comment:
Not sure that this is needed if you check the return code in the methods every time you perform a http request